src/TLuaInterpreter.cpp

```diff
@@ -5507,112 +5507,135 @@ int TLuaInterpreter::showToolBar(lua_State* L)
 // Documentation: https://wiki.mudlet.org/w/Manual:Lua_Functions#sendATCP
 int TLuaInterpreter::sendATCP(lua_State* L)
 {
+    Host& host = getHostFromLua(L);
     string msg;
     if (!lua_isstring(L, 1)) {
-        lua_pushstring(L, "sendATCP: what do you want to send?");
-        lua_error(L);
-        return 1;
+        lua_pushfstring(L, "sendATCP: bad argument #1 type (message as string expected, got %1!)", luaL_typename(L,1));
+        return lua_error(L);
     } else {
-        msg = lua_tostring(L, 1);
+        msg = host.mTelnet.encodeAndCookBytes(lua_tostring(L, 1));
     }
 
     string what;
-    if (!lua_isstring(L, 2)) {
-        lua_pushstring(L, "sendATCP: wrong argument type");
-        lua_error(L);
-        return 1;
-    } else {
-        what = lua_tostring(L, 2);
+    if (lua_gettop(L) > 1) {
+        if (!lua_isstring(L, 2)) {
+            lua_pushfstring(L, "sendATCP: bad argument #2 type (what as string is optional, got %1!)", luaL_typename(L,2));
+            return lua_error(L);
+        } else {
+            what = host.mTelnet.encodeAndCookBytes(lua_tostring(L, 2));
+        }
     }
-    string _h;
-    _h += TN_IAC;
-    _h += TN_SB;
-    _h += static_cast<char>(200);
-    _h += msg;
+
+    string output;
+    output += TN_IAC;
+    output += TN_SB;
+    output += OPT_ATCP;
+    output += msg;
     if (!what.empty()) {
-        _h += " ";
-        _h += what;
+        output += " ";
+        output += what;
     }
-    _h += TN_IAC;
-    _h += TN_SE;
+    output += TN_IAC;
+    output += TN_SE;
 
-    Host& host = getHostFromLua(L);
-    host.mTelnet.socketOutRaw(_h);
-    return 0;
+    if (!host.mTelnet.isATCPEnabled()) {
+        lua_pushnil(L);
+        lua_pushstring(L, "ATCP is not currently enabled by negotiation with the server");
```

Comment:
It ***hasn't*** been disabled, it just has not been negotiated to be *enabled* **which is not the same thing**. Saying "*the Server has disabled it*" is not correct - for instance we might have chosen to enable GMCP instead which then prevents Mudlet from negotiating to use it, even though the Server might be perfectly willing to use the older protocol! :nerd_face:

Reply:
OK, how about "ATCP is disabled in settings". The current message is still quite difficult to understand, especially for a non-fluent English speaker.

Reply:
I was trying to be vague about the cause because there could be multiple reasons - and that message covers all of them:

+ If, for instance, this is used in a script that is run just after the profile is loaded and the telnet options negotiation has not completed (can this happen BTW?) it might be that ATCP has not been negotiated yet but it still could be.

+ Alternatively it might have gotten disabled because, as I am aware, Mudlet will disable it if GMCP has been negotiated instead.

+ Also, the profile might be being used *off-line* - i.e. disconnected from the Server and this OOB protocol *may* then seem to work or not (though if a replay is in progress I am not sure how this sort of OOB would behave - assuming the scripts are the same as they were when the replay was made it might still seem to go through the same sequence of data packets even though there is no link between the data that is not transmitted to a real server and the faked data that seems to come from it that is the recording...)

The bottom line is that as far as a script is concerned the message is: "*You cannot do that with the way things are right now!*" and reading anything more into that needs more context...

Reply:
So would "ATCP is not currently enabled" work to cover all those cases?

The original message is just confusing, so I'm trying to find text that's clearer.

Reply:
:+1:  Yeah, all in all, I think that is an acceptable compromise.